// ts-mysql-parser/dist/lib/parsers-common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// Describes the type of a given query.
var MySQLQueryType;
(function (MySQLQueryType) {
    MySQLQueryType[MySQLQueryType["QtUnknown"] = 0] = "QtUnknown";
    MySQLQueryType[MySQLQueryType["QtAmbiguous"] = 1] = "QtAmbiguous";
    // DDL
    MySQLQueryType[MySQLQueryType["QtAlterDatabase"] = 2] = "QtAlterDatabase";
    MySQLQueryType[MySQLQueryType["QtAlterLogFileGroup"] = 3] = "QtAlterLogFileGroup";
    MySQLQueryType[MySQLQueryType["QtAlterFunction"] = 4] = "QtAlterFunction";
    MySQLQueryType[MySQLQueryType["QtAlterProcedure"] = 5] = "QtAlterProcedure";
    MySQLQueryType[MySQLQueryType["QtAlterServer"] = 6] = "QtAlterServer";
    MySQLQueryType[MySQLQueryType["QtAlterTable"] = 7] = "QtAlterTable";
    MySQLQueryType[MySQLQueryType["QtAlterTableSpace"] = 8] = "QtAlterTableSpace";
    MySQLQueryType[MySQLQueryType["QtAlterEvent"] = 9] = "QtAlterEvent";
    MySQLQueryType[MySQLQueryType["QtAlterView"] = 10] = "QtAlterView";
    MySQLQueryType[MySQLQueryType["QtCreateTable"] = 11] = "QtCreateTable";
    MySQLQueryType[MySQLQueryType["QtCreateIndex"] = 12] = "QtCreateIndex";
    MySQLQueryType[MySQLQueryType["QtCreateDatabase"] = 13] = "QtCreateDatabase";
    MySQLQueryType[MySQLQueryType["QtCreateEvent"] = 14] = "QtCreateEvent";
    MySQLQueryType[MySQLQueryType["QtCreateView"] = 15] = "QtCreateView";
    MySQLQueryType[MySQLQueryType["QtCreateRoutine"] = 16] = "QtCreateRoutine";
    MySQLQueryType[MySQLQueryType["QtCreateProcedure"] = 17] = "QtCreateProcedure";
    MySQLQueryType[MySQLQueryType["QtCreateFunction"] = 18] = "QtCreateFunction";
    MySQLQueryType[MySQLQueryType["QtCreateUdf"] = 19] = "QtCreateUdf";
    MySQLQueryType[MySQLQueryType["QtCreateTrigger"] = 20] = "QtCreateTrigger";
    MySQLQueryType[MySQLQueryType["QtCreateLogFileGroup"] = 21] = "QtCreateLogFileGroup";
    MySQLQueryType[MySQLQueryType["QtCreateServer"] = 22] = "QtCreateServer";
    MySQLQueryType[MySQLQueryType["QtCreateTableSpace"] = 23] = "QtCreateTableSpace";
    MySQLQueryType[MySQLQueryType["QtDropDatabase"] = 24] = "QtDropDatabase";
    MySQLQueryType[MySQLQueryType["QtDropEvent"] = 25] = "QtDropEvent";
    MySQLQueryType[MySQLQueryType["QtDropFunction"] = 26] = "QtDropFunction";
    MySQLQueryType[MySQLQueryType["QtDropProcedure"] = 27] = "QtDropProcedure";
    MySQLQueryType[MySQLQueryType["QtDropIndex"] = 28] = "QtDropIndex";
    MySQLQueryType[MySQLQueryType["QtDropLogfileGroup"] = 29] = "QtDropLogfileGroup";
    MySQLQueryType[MySQLQueryType["QtDropServer"] = 30] = "QtDropServer";
    MySQLQueryType[MySQLQueryType["QtDropTable"] = 31] = "QtDropTable";
    MySQLQueryType[MySQLQueryType["QtDropTablespace"] = 32] = "QtDropTablespace";
    MySQLQueryType[MySQLQueryType["QtDropTrigger"] = 33] = "QtDropTrigger";
    MySQLQueryType[MySQLQueryType["QtDropView"] = 34] = "QtDropView";
    MySQLQueryType[MySQLQueryType["QtRenameTable"] = 35] = "QtRenameTable";
    MySQLQueryType[MySQLQueryType["QtTruncateTable"] = 36] = "QtTruncateTable";
    // DML
    MySQLQueryType[MySQLQueryType["QtCall"] = 37] = "QtCall";
    MySQLQueryType[MySQLQueryType["QtDelete"] = 38] = "QtDelete";
    MySQLQueryType[MySQLQueryType["QtDo"] = 39] = "QtDo";
    MySQLQueryType[MySQLQueryType["QtHandler"] = 40] = "QtHandler";
    MySQLQueryType[MySQLQueryType["QtInsert"] = 41] = "QtInsert";
    MySQLQueryType[MySQLQueryType["QtLoadData"] = 42] = "QtLoadData";
    MySQLQueryType[MySQLQueryType["QtLoadXML"] = 43] = "QtLoadXML";
    MySQLQueryType[MySQLQueryType["QtReplace"] = 44] = "QtReplace";
    MySQLQueryType[MySQLQueryType["QtSelect"] = 45] = "QtSelect";
    MySQLQueryType[MySQLQueryType["QtUpdate"] = 46] = "QtUpdate";
    MySQLQueryType[MySQLQueryType["QtPartition"] = 47] = "QtPartition";
    MySQLQueryType[MySQLQueryType["QtStartTransaction"] = 48] = "QtStartTransaction";
    MySQLQueryType[MySQLQueryType["QtBeginWork"] = 49] = "QtBeginWork";
    MySQLQueryType[MySQLQueryType["QtCommit"] = 50] = "QtCommit";
    MySQLQueryType[MySQLQueryType["QtRollbackWork"] = 51] = "QtRollbackWork";
    MySQLQueryType[MySQLQueryType["QtSetAutoCommit"] = 52] = "QtSetAutoCommit";
    MySQLQueryType[MySQLQueryType["QtSetTransaction"] = 53] = "QtSetTransaction";
    MySQLQueryType[MySQLQueryType["QtSavepoint"] = 54] = "QtSavepoint";
    MySQLQueryType[MySQLQueryType["QtReleaseSavepoint"] = 55] = "QtReleaseSavepoint";
    MySQLQueryType[MySQLQueryType["QtRollbackSavepoint"] = 56] = "QtRollbackSavepoint";
    MySQLQueryType[MySQLQueryType["QtLock"] = 57] = "QtLock";
    MySQLQueryType[MySQLQueryType["QtUnlock"] = 58] = "QtUnlock";
    MySQLQueryType[MySQLQueryType["QtXA"] = 59] = "QtXA";
    MySQLQueryType[MySQLQueryType["QtPurge"] = 60] = "QtPurge";
    MySQLQueryType[MySQLQueryType["QtChangeMaster"] = 61] = "QtChangeMaster";
    MySQLQueryType[MySQLQueryType["QtReset"] = 62] = "QtReset";
    MySQLQueryType[MySQLQueryType["QtResetMaster"] = 63] = "QtResetMaster";
    MySQLQueryType[MySQLQueryType["QtResetSlave"] = 64] = "QtResetSlave";
    MySQLQueryType[MySQLQueryType["QtStartSlave"] = 65] = "QtStartSlave";
    MySQLQueryType[MySQLQueryType["QtStopSlave"] = 66] = "QtStopSlave";
    MySQLQueryType[MySQLQueryType["QtLoadDataMaster"] = 67] = "QtLoadDataMaster";
    MySQLQueryType[MySQLQueryType["QtLoadTableMaster"] = 68] = "QtLoadTableMaster";
    MySQLQueryType[MySQLQueryType["QtPrepare"] = 69] = "QtPrepare";
    MySQLQueryType[MySQLQueryType["QtExecute"] = 70] = "QtExecute";
    MySQLQueryType[MySQLQueryType["QtDeallocate"] = 71] = "QtDeallocate";
    // Database administration
    MySQLQueryType[MySQLQueryType["QtAlterUser"] = 72] = "QtAlterUser";
    MySQLQueryType[MySQLQueryType["QtCreateUser"] = 73] = "QtCreateUser";
    MySQLQueryType[MySQLQueryType["QtDropUser"] = 74] = "QtDropUser";
    MySQLQueryType[MySQLQueryType["QtGrantProxy"] = 75] = "QtGrantProxy";
    MySQLQueryType[MySQLQueryType["QtGrant"] = 76] = "QtGrant";
    MySQLQueryType[MySQLQueryType["QtRenameUser"] = 77] = "QtRenameUser";
    MySQLQueryType[MySQLQueryType["QtRevokeProxy"] = 78] = "QtRevokeProxy";
    MySQLQueryType[MySQLQueryType["QtRevoke"] = 79] = "QtRevoke";
    MySQLQueryType[MySQLQueryType["QtAnalyzeTable"] = 80] = "QtAnalyzeTable";
    MySQLQueryType[MySQLQueryType["QtCheckTable"] = 81] = "QtCheckTable";
    MySQLQueryType[MySQLQueryType["QtChecksumTable"] = 82] = "QtChecksumTable";
    MySQLQueryType[MySQLQueryType["QtOptimizeTable"] = 83] = "QtOptimizeTable";
    MySQLQueryType[MySQLQueryType["QtRepairTable"] = 84] = "QtRepairTable";
    MySQLQueryType[MySQLQueryType["QtBackUpTable"] = 85] = "QtBackUpTable";
    MySQLQueryType[MySQLQueryType["QtRestoreTable"] = 86] = "QtRestoreTable";
    MySQLQueryType[MySQLQueryType["QtInstallPlugin"] = 87] = "QtInstallPlugin";
    MySQLQueryType[MySQLQueryType["QtUninstallPlugin"] = 88] = "QtUninstallPlugin";
    MySQLQueryType[MySQLQueryType["QtSet"] = 89] = "QtSet";
    MySQLQueryType[MySQLQueryType["QtSetPassword"] = 90] = "QtSetPassword";
    MySQLQueryType[MySQLQueryType["QtShow"] = 91] = "QtShow";
    MySQLQueryType[MySQLQueryType["QtShowAuthors"] = 92] = "QtShowAuthors";
    MySQLQueryType[MySQLQueryType["QtShowBinaryLogs"] = 93] = "QtShowBinaryLogs";
    MySQLQueryType[MySQLQueryType["QtShowBinlogEvents"] = 94] = "QtShowBinlogEvents";
    MySQLQueryType[MySQLQueryType["QtShowRelaylogEvents"] = 95] = "QtShowRelaylogEvents";
    MySQLQueryType[MySQLQueryType["QtShowCharset"] = 96] = "QtShowCharset";
    MySQLQueryType[MySQLQueryType["QtShowCollation"] = 97] = "QtShowCollation";
    MySQLQueryType[MySQLQueryType["QtShowColumns"] = 98] = "QtShowColumns";
    MySQLQueryType[MySQLQueryType["QtShowContributors"] = 99] = "QtShowContributors";
    MySQLQueryType[MySQLQueryType["QtShowCreateDatabase"] = 100] = "QtShowCreateDatabase";
    MySQLQueryType[MySQLQueryType["QtShowCreateEvent"] = 101] = "QtShowCreateEvent";
    MySQLQueryType[MySQLQueryType["QtShowCreateFunction"] = 102] = "QtShowCreateFunction";
    MySQLQueryType[MySQLQueryType["QtShowCreateProcedure"] = 103] = "QtShowCreateProcedure";
    MySQLQueryType[MySQLQueryType["QtShowCreateTable"] = 104] = "QtShowCreateTable";
    MySQLQueryType[MySQLQueryType["QtShowCreateTrigger"] = 105] = "QtShowCreateTrigger";
    MySQLQueryType[MySQLQueryType["QtShowCreateView"] = 106] = "QtShowCreateView";
    MySQLQueryType[MySQLQueryType["QtShowDatabases"] = 107] = "QtShowDatabases";
    MySQLQueryType[MySQLQueryType["QtShowEngineStatus"] = 108] = "QtShowEngineStatus";
    MySQLQueryType[MySQLQueryType["QtShowStorageEngines"] = 109] = "QtShowStorageEngines";
    MySQLQueryType[MySQLQueryType["QtShowErrors"] = 110] = "QtShowErrors";
    MySQLQueryType[MySQLQueryType["QtShowEvents"] = 111] = "QtShowEvents";
    MySQLQueryType[MySQLQueryType["QtShowFunctionCode"] = 112] = "QtShowFunctionCode";
    MySQLQueryType[MySQLQueryType["QtShowFunctionStatus"] = 113] = "QtShowFunctionStatus";
    MySQLQueryType[MySQLQueryType["QtShowGrants"] = 114] = "QtShowGrants";
    MySQLQueryType[MySQLQueryType["QtShowIndexes"] = 115] = "QtShowIndexes";
    MySQLQueryType[MySQLQueryType["QtShowInnoDBStatus"] = 116] = "QtShowInnoDBStatus";
    MySQLQueryType[MySQLQueryType["QtShowMasterStatus"] = 117] = "QtShowMasterStatus";
    MySQLQueryType[MySQLQueryType["QtShowOpenTables"] = 118] = "QtShowOpenTables";
    MySQLQueryType[MySQLQueryType["QtShowPlugins"] = 119] = "QtShowPlugins";
    MySQLQueryType[MySQLQueryType["QtShowProcedureStatus"] = 120] = "QtShowProcedureStatus";
    MySQLQueryType[MySQLQueryType["QtShowProcedureCode"] = 121] = "QtShowProcedureCode";
    MySQLQueryType[MySQLQueryType["QtShowPrivileges"] = 122] = "QtShowPrivileges";
    MySQLQueryType[MySQLQueryType["QtShowProcessList"] = 123] = "QtShowProcessList";
    MySQLQueryType[MySQLQueryType["QtShowProfile"] = 124] = "QtShowProfile";
    MySQLQueryType[MySQLQueryType["QtShowProfiles"] = 125] = "QtShowProfiles";
    MySQLQueryType[MySQLQueryType["QtShowSlaveHosts"] = 126] = "QtShowSlaveHosts";
    MySQLQueryType[MySQLQueryType["QtShowSlaveStatus"] = 127] = "QtShowSlaveStatus";
    MySQLQueryType[MySQLQueryType["QtShowStatus"] = 128] = "QtShowStatus";
    MySQLQueryType[MySQLQueryType["QtShowVariables"] = 129] = "QtShowVariables";
    MySQLQueryType[MySQLQueryType["QtShowTableStatus"] = 130] = "QtShowTableStatus";
    MySQLQueryType[MySQLQueryType["QtShowTables"] = 131] = "QtShowTables";
    MySQLQueryType[MySQLQueryType["QtShowTriggers"] = 132] = "QtShowTriggers";
    MySQLQueryType[MySQLQueryType["QtShowWarnings"] = 133] = "QtShowWarnings";
    MySQLQueryType[MySQLQueryType["QtCacheIndex"] = 134] = "QtCacheIndex";
    MySQLQueryType[MySQLQueryType["QtFlush"] = 135] = "QtFlush";
    MySQLQueryType[MySQLQueryType["QtKill"] = 136] = "QtKill";
    MySQLQueryType[MySQLQueryType["QtLoadIndex"] = 137] = "QtLoadIndex";
    MySQLQueryType[MySQLQueryType["QtExplainTable"] = 138] = "QtExplainTable";
    MySQLQueryType[MySQLQueryType["QtExplainStatement"] = 139] = "QtExplainStatement";
    MySQLQueryType[MySQLQueryType["QtHelp"] = 140] = "QtHelp";
    MySQLQueryType[MySQLQueryType["QtUse"] = 141] = "QtUse";
    MySQLQueryType[MySQLQueryType["QtSentinel"] = 142] = "QtSentinel";
})(MySQLQueryType = exports.MySQLQueryType || (exports.MySQLQueryType = {}));
