// ts-mysql-parser/dist/lib/rule-name.js
"use strict";
/* eslint-disable @typescript-eslint/camelcase */
Object.defineProperty(exports, "__esModule", { value: true });
var RuleName;
(function (RuleName) {
    RuleName["query"] = "query";
    RuleName["simpleStatement"] = "simpleStatement";
    RuleName["alterStatement"] = "alterStatement";
    RuleName["alterDatabase"] = "alterDatabase";
    RuleName["alterEvent"] = "alterEvent";
    RuleName["alterLogfileGroup"] = "alterLogfileGroup";
    RuleName["alterLogfileGroupOptions"] = "alterLogfileGroupOptions";
    RuleName["alterLogfileGroupOption"] = "alterLogfileGroupOption";
    RuleName["alterServer"] = "alterServer";
    RuleName["alterTable"] = "alterTable";
    RuleName["alterTableActions"] = "alterTableActions";
    RuleName["alterCommandList"] = "alterCommandList";
    RuleName["alterCommandsModifierList"] = "alterCommandsModifierList";
    RuleName["standaloneAlterCommands"] = "standaloneAlterCommands";
    RuleName["alterPartition"] = "alterPartition";
    RuleName["alterList"] = "alterList";
    RuleName["alterCommandsModifier"] = "alterCommandsModifier";
    RuleName["alterListItem"] = "alterListItem";
    RuleName["place"] = "place";
    RuleName["restrict"] = "restrict";
    RuleName["alterOrderList"] = "alterOrderList";
    RuleName["alterAlgorithmOption"] = "alterAlgorithmOption";
    RuleName["alterLockOption"] = "alterLockOption";
    RuleName["indexLockAndAlgorithm"] = "indexLockAndAlgorithm";
    RuleName["withValidation"] = "withValidation";
    RuleName["removePartitioning"] = "removePartitioning";
    RuleName["allOrPartitionNameList"] = "allOrPartitionNameList";
    RuleName["reorgPartitionRule"] = "reorgPartitionRule";
    RuleName["alterTablespace"] = "alterTablespace";
    RuleName["alterUndoTablespace"] = "alterUndoTablespace";
    RuleName["undoTableSpaceOptions"] = "undoTableSpaceOptions";
    RuleName["undoTableSpaceOption"] = "undoTableSpaceOption";
    RuleName["alterTablespaceOptions"] = "alterTablespaceOptions";
    RuleName["alterTablespaceOption"] = "alterTablespaceOption";
    RuleName["changeTablespaceOption"] = "changeTablespaceOption";
    RuleName["alterView"] = "alterView";
    RuleName["viewTail"] = "viewTail";
    RuleName["viewSelect"] = "viewSelect";
    RuleName["viewCheckOption"] = "viewCheckOption";
    RuleName["createStatement"] = "createStatement";
    RuleName["createDatabase"] = "createDatabase";
    RuleName["createDatabaseOption"] = "createDatabaseOption";
    RuleName["createTable"] = "createTable";
    RuleName["tableElementList"] = "tableElementList";
    RuleName["tableElement"] = "tableElement";
    RuleName["duplicateAsQueryExpression"] = "duplicateAsQueryExpression";
    RuleName["queryExpressionOrParens"] = "queryExpressionOrParens";
    RuleName["createRoutine"] = "createRoutine";
    RuleName["createProcedure"] = "createProcedure";
    RuleName["createFunction"] = "createFunction";
    RuleName["createUdf"] = "createUdf";
    RuleName["routineCreateOption"] = "routineCreateOption";
    RuleName["routineAlterOptions"] = "routineAlterOptions";
    RuleName["routineOption"] = "routineOption";
    RuleName["createIndex"] = "createIndex";
    RuleName["indexNameAndType"] = "indexNameAndType";
    RuleName["createIndexTarget"] = "createIndexTarget";
    RuleName["createLogfileGroup"] = "createLogfileGroup";
    RuleName["logfileGroupOptions"] = "logfileGroupOptions";
    RuleName["logfileGroupOption"] = "logfileGroupOption";
    RuleName["createServer"] = "createServer";
    RuleName["serverOptions"] = "serverOptions";
    RuleName["serverOption"] = "serverOption";
    RuleName["createTablespace"] = "createTablespace";
    RuleName["createUndoTablespace"] = "createUndoTablespace";
    RuleName["tsDataFileName"] = "tsDataFileName";
    RuleName["tsDataFile"] = "tsDataFile";
    RuleName["tablespaceOptions"] = "tablespaceOptions";
    RuleName["tablespaceOption"] = "tablespaceOption";
    RuleName["tsOptionInitialSize"] = "tsOptionInitialSize";
    RuleName["tsOptionUndoRedoBufferSize"] = "tsOptionUndoRedoBufferSize";
    RuleName["tsOptionAutoextendSize"] = "tsOptionAutoextendSize";
    RuleName["tsOptionMaxSize"] = "tsOptionMaxSize";
    RuleName["tsOptionExtentSize"] = "tsOptionExtentSize";
    RuleName["tsOptionNodegroup"] = "tsOptionNodegroup";
    RuleName["tsOptionEngine"] = "tsOptionEngine";
    RuleName["tsOptionWait"] = "tsOptionWait";
    RuleName["tsOptionComment"] = "tsOptionComment";
    RuleName["tsOptionFileblockSize"] = "tsOptionFileblockSize";
    RuleName["tsOptionEncryption"] = "tsOptionEncryption";
    RuleName["createView"] = "createView";
    RuleName["viewReplaceOrAlgorithm"] = "viewReplaceOrAlgorithm";
    RuleName["viewAlgorithm"] = "viewAlgorithm";
    RuleName["viewSuid"] = "viewSuid";
    RuleName["createTrigger"] = "createTrigger";
    RuleName["triggerFollowsPrecedesClause"] = "triggerFollowsPrecedesClause";
    RuleName["createEvent"] = "createEvent";
    RuleName["createRole"] = "createRole";
    RuleName["createSpatialReference"] = "createSpatialReference";
    RuleName["srsAttribute"] = "srsAttribute";
    RuleName["dropStatement"] = "dropStatement";
    RuleName["dropDatabase"] = "dropDatabase";
    RuleName["dropEvent"] = "dropEvent";
    RuleName["dropFunction"] = "dropFunction";
    RuleName["dropProcedure"] = "dropProcedure";
    RuleName["dropIndex"] = "dropIndex";
    RuleName["dropLogfileGroup"] = "dropLogfileGroup";
    RuleName["dropLogfileGroupOption"] = "dropLogfileGroupOption";
    RuleName["dropServer"] = "dropServer";
    RuleName["dropTable"] = "dropTable";
    RuleName["dropTableSpace"] = "dropTableSpace";
    RuleName["dropTrigger"] = "dropTrigger";
    RuleName["dropView"] = "dropView";
    RuleName["dropRole"] = "dropRole";
    RuleName["dropSpatialReference"] = "dropSpatialReference";
    RuleName["dropUndoTablespace"] = "dropUndoTablespace";
    RuleName["renameTableStatement"] = "renameTableStatement";
    RuleName["renamePair"] = "renamePair";
    RuleName["truncateTableStatement"] = "truncateTableStatement";
    RuleName["importStatement"] = "importStatement";
    RuleName["callStatement"] = "callStatement";
    RuleName["deleteStatement"] = "deleteStatement";
    RuleName["partitionDelete"] = "partitionDelete";
    RuleName["deleteStatementOption"] = "deleteStatementOption";
    RuleName["doStatement"] = "doStatement";
    RuleName["handlerStatement"] = "handlerStatement";
    RuleName["handlerReadOrScan"] = "handlerReadOrScan";
    RuleName["insertStatement"] = "insertStatement";
    RuleName["insertLockOption"] = "insertLockOption";
    RuleName["insertFromConstructor"] = "insertFromConstructor";
    RuleName["fields"] = "fields";
    RuleName["insertValues"] = "insertValues";
    RuleName["insertQueryExpression"] = "insertQueryExpression";
    RuleName["valueList"] = "valueList";
    RuleName["values"] = "values";
    RuleName["valuesReference"] = "valuesReference";
    RuleName["insertUpdateList"] = "insertUpdateList";
    RuleName["loadStatement"] = "loadStatement";
    RuleName["dataOrXml"] = "dataOrXml";
    RuleName["xmlRowsIdentifiedBy"] = "xmlRowsIdentifiedBy";
    RuleName["loadDataFileTail"] = "loadDataFileTail";
    RuleName["loadDataFileTargetList"] = "loadDataFileTargetList";
    RuleName["fieldOrVariableList"] = "fieldOrVariableList";
    RuleName["replaceStatement"] = "replaceStatement";
    RuleName["selectStatement"] = "selectStatement";
    RuleName["selectStatementWithInto"] = "selectStatementWithInto";
    RuleName["queryExpression"] = "queryExpression";
    RuleName["queryExpressionBody"] = "queryExpressionBody";
    RuleName["queryExpressionParens"] = "queryExpressionParens";
    RuleName["querySpecification"] = "querySpecification";
    RuleName["subquery"] = "subquery";
    RuleName["querySpecOption"] = "querySpecOption";
    RuleName["limitClause"] = "limitClause";
    RuleName["simpleLimitClause"] = "simpleLimitClause";
    RuleName["limitOptions"] = "limitOptions";
    RuleName["limitOption"] = "limitOption";
    RuleName["intoClause"] = "intoClause";
    RuleName["procedureAnalyseClause"] = "procedureAnalyseClause";
    RuleName["havingClause"] = "havingClause";
    RuleName["windowClause"] = "windowClause";
    RuleName["windowDefinition"] = "windowDefinition";
    RuleName["windowSpec"] = "windowSpec";
    RuleName["windowSpecDetails"] = "windowSpecDetails";
    RuleName["windowFrameClause"] = "windowFrameClause";
    RuleName["windowFrameUnits"] = "windowFrameUnits";
    RuleName["windowFrameExtent"] = "windowFrameExtent";
    RuleName["windowFrameStart"] = "windowFrameStart";
    RuleName["windowFrameBetween"] = "windowFrameBetween";
    RuleName["windowFrameBound"] = "windowFrameBound";
    RuleName["windowFrameExclusion"] = "windowFrameExclusion";
    RuleName["withClause"] = "withClause";
    RuleName["commonTableExpression"] = "commonTableExpression";
    RuleName["groupByClause"] = "groupByClause";
    RuleName["olapOption"] = "olapOption";
    RuleName["orderClause"] = "orderClause";
    RuleName["direction"] = "direction";
    RuleName["fromClause"] = "fromClause";
    RuleName["tableReferenceList"] = "tableReferenceList";
    RuleName["selectOption"] = "selectOption";
    RuleName["lockingClause"] = "lockingClause";
    RuleName["lockStrengh"] = "lockStrengh";
    RuleName["lockedRowAction"] = "lockedRowAction";
    RuleName["selectItemList"] = "selectItemList";
    RuleName["selectItem"] = "selectItem";
    RuleName["selectAlias"] = "selectAlias";
    RuleName["whereClause"] = "whereClause";
    RuleName["tableReference"] = "tableReference";
    RuleName["escapedTableReference"] = "escapedTableReference";
    RuleName["joinedTable"] = "joinedTable";
    RuleName["naturalJoinType"] = "naturalJoinType";
    RuleName["innerJoinType"] = "innerJoinType";
    RuleName["outerJoinType"] = "outerJoinType";
    RuleName["tableFactor"] = "tableFactor";
    RuleName["singleTable"] = "singleTable";
    RuleName["singleTableParens"] = "singleTableParens";
    RuleName["derivedTable"] = "derivedTable";
    RuleName["tableReferenceListParens"] = "tableReferenceListParens";
    RuleName["tableFunction"] = "tableFunction";
    RuleName["columnsClause"] = "columnsClause";
    RuleName["jtColumn"] = "jtColumn";
    RuleName["onEmptyOrError"] = "onEmptyOrError";
    RuleName["onEmpty"] = "onEmpty";
    RuleName["onError"] = "onError";
    RuleName["jtOnResponse"] = "jtOnResponse";
    RuleName["unionOption"] = "unionOption";
    RuleName["tableAlias"] = "tableAlias";
    RuleName["indexHintList"] = "indexHintList";
    RuleName["indexHint"] = "indexHint";
    RuleName["indexHintType"] = "indexHintType";
    RuleName["keyOrIndex"] = "keyOrIndex";
    RuleName["constraintKeyType"] = "constraintKeyType";
    RuleName["indexHintClause"] = "indexHintClause";
    RuleName["indexList"] = "indexList";
    RuleName["indexListElement"] = "indexListElement";
    RuleName["updateStatement"] = "updateStatement";
    RuleName["transactionOrLockingStatement"] = "transactionOrLockingStatement";
    RuleName["transactionStatement"] = "transactionStatement";
    RuleName["beginWork"] = "beginWork";
    RuleName["transactionCharacteristic"] = "transactionCharacteristic";
    RuleName["savepointStatement"] = "savepointStatement";
    RuleName["lockStatement"] = "lockStatement";
    RuleName["lockItem"] = "lockItem";
    RuleName["lockOption"] = "lockOption";
    RuleName["xaStatement"] = "xaStatement";
    RuleName["xaConvert"] = "xaConvert";
    RuleName["xid"] = "xid";
    RuleName["replicationStatement"] = "replicationStatement";
    RuleName["resetOption"] = "resetOption";
    RuleName["masterResetOptions"] = "masterResetOptions";
    RuleName["replicationLoad"] = "replicationLoad";
    RuleName["changeMaster"] = "changeMaster";
    RuleName["changeMasterOptions"] = "changeMasterOptions";
    RuleName["masterOption"] = "masterOption";
    RuleName["privilegeCheckDef"] = "privilegeCheckDef";
    RuleName["masterTlsCiphersuitesDef"] = "masterTlsCiphersuitesDef";
    RuleName["masterFileDef"] = "masterFileDef";
    RuleName["serverIdList"] = "serverIdList";
    RuleName["changeReplication"] = "changeReplication";
    RuleName["filterDefinition"] = "filterDefinition";
    RuleName["filterDbList"] = "filterDbList";
    RuleName["filterTableList"] = "filterTableList";
    RuleName["filterStringList"] = "filterStringList";
    RuleName["filterWildDbTableString"] = "filterWildDbTableString";
    RuleName["filterDbPairList"] = "filterDbPairList";
    RuleName["slave"] = "slave";
    RuleName["slaveUntilOptions"] = "slaveUntilOptions";
    RuleName["slaveConnectionOptions"] = "slaveConnectionOptions";
    RuleName["slaveThreadOptions"] = "slaveThreadOptions";
    RuleName["slaveThreadOption"] = "slaveThreadOption";
    RuleName["groupReplication"] = "groupReplication";
    RuleName["preparedStatement"] = "preparedStatement";
    RuleName["executeStatement"] = "executeStatement";
    RuleName["executeVarList"] = "executeVarList";
    RuleName["cloneStatement"] = "cloneStatement";
    RuleName["dataDirSSL"] = "dataDirSSL";
    RuleName["ssl"] = "ssl";
    RuleName["accountManagementStatement"] = "accountManagementStatement";
    RuleName["alterUser"] = "alterUser";
    RuleName["alterUserTail"] = "alterUserTail";
    RuleName["userFunction"] = "userFunction";
    RuleName["createUser"] = "createUser";
    RuleName["createUserTail"] = "createUserTail";
    RuleName["defaultRoleClause"] = "defaultRoleClause";
    RuleName["requireClause"] = "requireClause";
    RuleName["connectOptions"] = "connectOptions";
    RuleName["accountLockPasswordExpireOptions"] = "accountLockPasswordExpireOptions";
    RuleName["dropUser"] = "dropUser";
    RuleName["grant"] = "grant";
    RuleName["grantTargetList"] = "grantTargetList";
    RuleName["grantOptions"] = "grantOptions";
    RuleName["exceptRoleList"] = "exceptRoleList";
    RuleName["withRoles"] = "withRoles";
    RuleName["grantAs"] = "grantAs";
    RuleName["versionedRequireClause"] = "versionedRequireClause";
    RuleName["renameUser"] = "renameUser";
    RuleName["revoke"] = "revoke";
    RuleName["onTypeTo"] = "onTypeTo";
    RuleName["aclType"] = "aclType";
    RuleName["roleOrPrivilegesList"] = "roleOrPrivilegesList";
    RuleName["roleOrPrivilege"] = "roleOrPrivilege";
    RuleName["grantIdentifier"] = "grantIdentifier";
    RuleName["requireList"] = "requireList";
    RuleName["requireListElement"] = "requireListElement";
    RuleName["grantOption"] = "grantOption";
    RuleName["setRole"] = "setRole";
    RuleName["roleList"] = "roleList";
    RuleName["role"] = "role";
    RuleName["tableAdministrationStatement"] = "tableAdministrationStatement";
    RuleName["histogram"] = "histogram";
    RuleName["checkOption"] = "checkOption";
    RuleName["repairType"] = "repairType";
    RuleName["installUninstallStatment"] = "installUninstallStatment";
    RuleName["setStatement"] = "setStatement";
    RuleName["startOptionValueList"] = "startOptionValueList";
    RuleName["transactionCharacteristics"] = "transactionCharacteristics";
    RuleName["transactionAccessMode"] = "transactionAccessMode";
    RuleName["isolationLevel"] = "isolationLevel";
    RuleName["optionValueListContinued"] = "optionValueListContinued";
    RuleName["optionValueNoOptionType"] = "optionValueNoOptionType";
    RuleName["optionValue"] = "optionValue";
    RuleName["setSystemVariable"] = "setSystemVariable";
    RuleName["startOptionValueListFollowingOptionType"] = "startOptionValueListFollowingOptionType";
    RuleName["optionValueFollowingOptionType"] = "optionValueFollowingOptionType";
    RuleName["setExprOrDefault"] = "setExprOrDefault";
    RuleName["showStatement"] = "showStatement";
    RuleName["showCommandType"] = "showCommandType";
    RuleName["nonBlocking"] = "nonBlocking";
    RuleName["fromOrIn"] = "fromOrIn";
    RuleName["inDb"] = "inDb";
    RuleName["profileType"] = "profileType";
    RuleName["otherAdministrativeStatement"] = "otherAdministrativeStatement";
    RuleName["keyCacheListOrParts"] = "keyCacheListOrParts";
    RuleName["keyCacheList"] = "keyCacheList";
    RuleName["assignToKeycache"] = "assignToKeycache";
    RuleName["assignToKeycachePartition"] = "assignToKeycachePartition";
    RuleName["cacheKeyList"] = "cacheKeyList";
    RuleName["keyUsageElement"] = "keyUsageElement";
    RuleName["keyUsageList"] = "keyUsageList";
    RuleName["flushOption"] = "flushOption";
    RuleName["logType"] = "logType";
    RuleName["flushTables"] = "flushTables";
    RuleName["flushTablesOptions"] = "flushTablesOptions";
    RuleName["preloadTail"] = "preloadTail";
    RuleName["preloadList"] = "preloadList";
    RuleName["preloadKeys"] = "preloadKeys";
    RuleName["adminPartition"] = "adminPartition";
    RuleName["resourceGroupManagement"] = "resourceGroupManagement";
    RuleName["createResourceGroup"] = "createResourceGroup";
    RuleName["resourceGroupVcpuList"] = "resourceGroupVcpuList";
    RuleName["vcpuNumOrRange"] = "vcpuNumOrRange";
    RuleName["resourceGroupPriority"] = "resourceGroupPriority";
    RuleName["resourceGroupEnableDisable"] = "resourceGroupEnableDisable";
    RuleName["alterResourceGroup"] = "alterResourceGroup";
    RuleName["setResourceGroup"] = "setResourceGroup";
    RuleName["threadIdList"] = "threadIdList";
    RuleName["dropResourceGroup"] = "dropResourceGroup";
    RuleName["utilityStatement"] = "utilityStatement";
    RuleName["describeCommand"] = "describeCommand";
    RuleName["explainCommand"] = "explainCommand";
    RuleName["explainableStatement"] = "explainableStatement";
    RuleName["helpCommand"] = "helpCommand";
    RuleName["useCommand"] = "useCommand";
    RuleName["restartServer"] = "restartServer";
    RuleName["expr"] = "expr";
    RuleName["boolPri"] = "boolPri";
    RuleName["compOp"] = "compOp";
    RuleName["predicate"] = "predicate";
    RuleName["predicateOperations"] = "predicateOperations";
    RuleName["bitExpr"] = "bitExpr";
    RuleName["simpleExpr"] = "simpleExpr";
    RuleName["arrayCast"] = "arrayCast";
    RuleName["jsonOperator"] = "jsonOperator";
    RuleName["sumExpr"] = "sumExpr";
    RuleName["groupingOperation"] = "groupingOperation";
    RuleName["windowFunctionCall"] = "windowFunctionCall";
    RuleName["windowingClause"] = "windowingClause";
    RuleName["leadLagInfo"] = "leadLagInfo";
    RuleName["nullTreatment"] = "nullTreatment";
    RuleName["jsonFunction"] = "jsonFunction";
    RuleName["inSumExpr"] = "inSumExpr";
    RuleName["identListArg"] = "identListArg";
    RuleName["identList"] = "identList";
    RuleName["fulltextOptions"] = "fulltextOptions";
    RuleName["runtimeFunctionCall"] = "runtimeFunctionCall";
    RuleName["geometryFunction"] = "geometryFunction";
    RuleName["timeFunctionParameters"] = "timeFunctionParameters";
    RuleName["fractionalPrecision"] = "fractionalPrecision";
    RuleName["weightStringLevels"] = "weightStringLevels";
    RuleName["weightStringLevelListItem"] = "weightStringLevelListItem";
    RuleName["dateTimeTtype"] = "dateTimeTtype";
    RuleName["trimFunction"] = "trimFunction";
    RuleName["substringFunction"] = "substringFunction";
    RuleName["functionCall"] = "functionCall";
    RuleName["udfExprList"] = "udfExprList";
    RuleName["udfExpr"] = "udfExpr";
    RuleName["variable"] = "variable";
    RuleName["userVariable"] = "userVariable";
    RuleName["systemVariable"] = "systemVariable";
    RuleName["internalVariableName"] = "internalVariableName";
    RuleName["whenExpression"] = "whenExpression";
    RuleName["thenExpression"] = "thenExpression";
    RuleName["elseExpression"] = "elseExpression";
    RuleName["castType"] = "castType";
    RuleName["exprList"] = "exprList";
    RuleName["charset"] = "charset";
    RuleName["notRule"] = "notRule";
    RuleName["not2Rule"] = "not2Rule";
    RuleName["interval"] = "interval";
    RuleName["intervalTimeStamp"] = "intervalTimeStamp";
    RuleName["exprListWithParentheses"] = "exprListWithParentheses";
    RuleName["exprWithParentheses"] = "exprWithParentheses";
    RuleName["simpleExprWithParentheses"] = "simpleExprWithParentheses";
    RuleName["orderList"] = "orderList";
    RuleName["orderExpression"] = "orderExpression";
    RuleName["groupList"] = "groupList";
    RuleName["groupingExpression"] = "groupingExpression";
    RuleName["channel"] = "channel";
    RuleName["compoundStatement"] = "compoundStatement";
    RuleName["returnStatement"] = "returnStatement";
    RuleName["ifStatement"] = "ifStatement";
    RuleName["ifBody"] = "ifBody";
    RuleName["thenStatement"] = "thenStatement";
    RuleName["compoundStatementList"] = "compoundStatementList";
    RuleName["caseStatement"] = "caseStatement";
    RuleName["elseStatement"] = "elseStatement";
    RuleName["labeledBlock"] = "labeledBlock";
    RuleName["unlabeledBlock"] = "unlabeledBlock";
    RuleName["label"] = "label";
    RuleName["beginEndBlock"] = "beginEndBlock";
    RuleName["labeledControl"] = "labeledControl";
    RuleName["unlabeledControl"] = "unlabeledControl";
    RuleName["loopBlock"] = "loopBlock";
    RuleName["whileDoBlock"] = "whileDoBlock";
    RuleName["repeatUntilBlock"] = "repeatUntilBlock";
    RuleName["spDeclarations"] = "spDeclarations";
    RuleName["spDeclaration"] = "spDeclaration";
    RuleName["variableDeclaration"] = "variableDeclaration";
    RuleName["conditionDeclaration"] = "conditionDeclaration";
    RuleName["spCondition"] = "spCondition";
    RuleName["sqlstate"] = "sqlstate";
    RuleName["handlerDeclaration"] = "handlerDeclaration";
    RuleName["handlerCondition"] = "handlerCondition";
    RuleName["cursorDeclaration"] = "cursorDeclaration";
    RuleName["iterateStatement"] = "iterateStatement";
    RuleName["leaveStatement"] = "leaveStatement";
    RuleName["getDiagnostics"] = "getDiagnostics";
    RuleName["signalAllowedExpr"] = "signalAllowedExpr";
    RuleName["statementInformationItem"] = "statementInformationItem";
    RuleName["conditionInformationItem"] = "conditionInformationItem";
    RuleName["signalInformationItemName"] = "signalInformationItemName";
    RuleName["signalStatement"] = "signalStatement";
    RuleName["resignalStatement"] = "resignalStatement";
    RuleName["signalInformationItem"] = "signalInformationItem";
    RuleName["cursorOpen"] = "cursorOpen";
    RuleName["cursorClose"] = "cursorClose";
    RuleName["cursorFetch"] = "cursorFetch";
    RuleName["schedule"] = "schedule";
    RuleName["columnDefinition"] = "columnDefinition";
    RuleName["checkOrReferences"] = "checkOrReferences";
    RuleName["checkConstraint"] = "checkConstraint";
    RuleName["constraintEnforcement"] = "constraintEnforcement";
    RuleName["tableConstraintDef"] = "tableConstraintDef";
    RuleName["constraintName"] = "constraintName";
    RuleName["fieldDefinition"] = "fieldDefinition";
    RuleName["columnAttribute"] = "columnAttribute";
    RuleName["columnFormat"] = "columnFormat";
    RuleName["storageMedia"] = "storageMedia";
    RuleName["gcolAttribute"] = "gcolAttribute";
    RuleName["references"] = "references";
    RuleName["deleteOption"] = "deleteOption";
    RuleName["keyList"] = "keyList";
    RuleName["keyPart"] = "keyPart";
    RuleName["keyListWithExpression"] = "keyListWithExpression";
    RuleName["keyPartOrExpression"] = "keyPartOrExpression";
    RuleName["keyListVariants"] = "keyListVariants";
    RuleName["indexType"] = "indexType";
    RuleName["indexOption"] = "indexOption";
    RuleName["commonIndexOption"] = "commonIndexOption";
    RuleName["visibility"] = "visibility";
    RuleName["indexTypeClause"] = "indexTypeClause";
    RuleName["fulltextIndexOption"] = "fulltextIndexOption";
    RuleName["spatialIndexOption"] = "spatialIndexOption";
    RuleName["dataTypeDefinition"] = "dataTypeDefinition";
    RuleName["dataType"] = "dataType";
    RuleName["nchar"] = "nchar";
    RuleName["realType"] = "realType";
    RuleName["fieldLength"] = "fieldLength";
    RuleName["fieldOptions"] = "fieldOptions";
    RuleName["charsetWithOptBinary"] = "charsetWithOptBinary";
    RuleName["ascii"] = "ascii";
    RuleName["unicode"] = "unicode";
    RuleName["wsNumCodepoints"] = "wsNumCodepoints";
    RuleName["typeDatetimePrecision"] = "typeDatetimePrecision";
    RuleName["charsetName"] = "charsetName";
    RuleName["collationName"] = "collationName";
    RuleName["createTableOptions"] = "createTableOptions";
    RuleName["createTableOptionsSpaceSeparated"] = "createTableOptionsSpaceSeparated";
    RuleName["createTableOption"] = "createTableOption";
    RuleName["ternaryOption"] = "ternaryOption";
    RuleName["defaultCollation"] = "defaultCollation";
    RuleName["defaultEncryption"] = "defaultEncryption";
    RuleName["defaultCharset"] = "defaultCharset";
    RuleName["partitionClause"] = "partitionClause";
    RuleName["partitionTypeDef"] = "partitionTypeDef";
    RuleName["subPartitions"] = "subPartitions";
    RuleName["partitionKeyAlgorithm"] = "partitionKeyAlgorithm";
    RuleName["partitionDefinitions"] = "partitionDefinitions";
    RuleName["partitionDefinition"] = "partitionDefinition";
    RuleName["partitionValuesIn"] = "partitionValuesIn";
    RuleName["partitionOption"] = "partitionOption";
    RuleName["subpartitionDefinition"] = "subpartitionDefinition";
    RuleName["partitionValueItemListParen"] = "partitionValueItemListParen";
    RuleName["partitionValueItem"] = "partitionValueItem";
    RuleName["definerClause"] = "definerClause";
    RuleName["ifExists"] = "ifExists";
    RuleName["ifNotExists"] = "ifNotExists";
    RuleName["procedureParameter"] = "procedureParameter";
    RuleName["functionParameter"] = "functionParameter";
    RuleName["collate"] = "collate";
    RuleName["typeWithOptCollate"] = "typeWithOptCollate";
    RuleName["schemaIdentifierPair"] = "schemaIdentifierPair";
    RuleName["viewRefList"] = "viewRefList";
    RuleName["updateList"] = "updateList";
    RuleName["updateElement"] = "updateElement";
    RuleName["charsetClause"] = "charsetClause";
    RuleName["fieldsClause"] = "fieldsClause";
    RuleName["fieldTerm"] = "fieldTerm";
    RuleName["linesClause"] = "linesClause";
    RuleName["lineTerm"] = "lineTerm";
    RuleName["userList"] = "userList";
    RuleName["createUserList"] = "createUserList";
    RuleName["alterUserList"] = "alterUserList";
    RuleName["createUserEntry"] = "createUserEntry";
    RuleName["alterUserEntry"] = "alterUserEntry";
    RuleName["retainCurrentPassword"] = "retainCurrentPassword";
    RuleName["discardOldPassword"] = "discardOldPassword";
    RuleName["replacePassword"] = "replacePassword";
    RuleName["userIdentifierOrText"] = "userIdentifierOrText";
    RuleName["user"] = "user";
    RuleName["likeClause"] = "likeClause";
    RuleName["likeOrWhere"] = "likeOrWhere";
    RuleName["onlineOption"] = "onlineOption";
    RuleName["noWriteToBinLog"] = "noWriteToBinLog";
    RuleName["usePartition"] = "usePartition";
    RuleName["fieldIdentifier"] = "fieldIdentifier";
    RuleName["columnName"] = "columnName";
    RuleName["columnInternalRef"] = "columnInternalRef";
    RuleName["columnInternalRefList"] = "columnInternalRefList";
    RuleName["columnRef"] = "columnRef";
    RuleName["insertIdentifier"] = "insertIdentifier";
    RuleName["indexName"] = "indexName";
    RuleName["indexRef"] = "indexRef";
    RuleName["tableWild"] = "tableWild";
    RuleName["schemaName"] = "schemaName";
    RuleName["schemaRef"] = "schemaRef";
    RuleName["procedureName"] = "procedureName";
    RuleName["procedureRef"] = "procedureRef";
    RuleName["functionName"] = "functionName";
    RuleName["functionRef"] = "functionRef";
    RuleName["triggerName"] = "triggerName";
    RuleName["triggerRef"] = "triggerRef";
    RuleName["viewName"] = "viewName";
    RuleName["viewRef"] = "viewRef";
    RuleName["tablespaceName"] = "tablespaceName";
    RuleName["tablespaceRef"] = "tablespaceRef";
    RuleName["logfileGroupName"] = "logfileGroupName";
    RuleName["logfileGroupRef"] = "logfileGroupRef";
    RuleName["eventName"] = "eventName";
    RuleName["eventRef"] = "eventRef";
    RuleName["udfName"] = "udfName";
    RuleName["serverName"] = "serverName";
    RuleName["serverRef"] = "serverRef";
    RuleName["engineRef"] = "engineRef";
    RuleName["tableName"] = "tableName";
    RuleName["filterTableRef"] = "filterTableRef";
    RuleName["tableRefWithWildcard"] = "tableRefWithWildcard";
    RuleName["tableRef"] = "tableRef";
    RuleName["tableRefList"] = "tableRefList";
    RuleName["tableAliasRefList"] = "tableAliasRefList";
    RuleName["parameterName"] = "parameterName";
    RuleName["labelIdentifier"] = "labelIdentifier";
    RuleName["labelRef"] = "labelRef";
    RuleName["roleIdentifier"] = "roleIdentifier";
    RuleName["roleRef"] = "roleRef";
    RuleName["pluginRef"] = "pluginRef";
    RuleName["componentRef"] = "componentRef";
    RuleName["resourceGroupRef"] = "resourceGroupRef";
    RuleName["windowName"] = "windowName";
    RuleName["pureIdentifier"] = "pureIdentifier";
    RuleName["identifier"] = "identifier";
    RuleName["identifierList"] = "identifierList";
    RuleName["identifierListWithParentheses"] = "identifierListWithParentheses";
    RuleName["qualifiedIdentifier"] = "qualifiedIdentifier";
    RuleName["simpleIdentifier"] = "simpleIdentifier";
    RuleName["dotIdentifier"] = "dotIdentifier";
    RuleName["ulong_number"] = "ulong_number";
    RuleName["real_ulong_number"] = "real_ulong_number";
    RuleName["ulonglong_number"] = "ulonglong_number";
    RuleName["real_ulonglong_number"] = "real_ulonglong_number";
    RuleName["literal"] = "literal";
    RuleName["signedLiteral"] = "signedLiteral";
    RuleName["stringList"] = "stringList";
    RuleName["textStringLiteral"] = "textStringLiteral";
    RuleName["textString"] = "textString";
    RuleName["textStringHash"] = "textStringHash";
    RuleName["textLiteral"] = "textLiteral";
    RuleName["textStringNoLinebreak"] = "textStringNoLinebreak";
    RuleName["textStringLiteralList"] = "textStringLiteralList";
    RuleName["numLiteral"] = "numLiteral";
    RuleName["boolLiteral"] = "boolLiteral";
    RuleName["nullLiteral"] = "nullLiteral";
    RuleName["temporalLiteral"] = "temporalLiteral";
    RuleName["floatOptions"] = "floatOptions";
    RuleName["standardFloatOptions"] = "standardFloatOptions";
    RuleName["precision"] = "precision";
    RuleName["textOrIdentifier"] = "textOrIdentifier";
    RuleName["lValueIdentifier"] = "lValueIdentifier";
    RuleName["roleIdentifierOrText"] = "roleIdentifierOrText";
    RuleName["sizeNumber"] = "sizeNumber";
    RuleName["parentheses"] = "parentheses";
    RuleName["equal"] = "equal";
    RuleName["optionType"] = "optionType";
    RuleName["varIdentType"] = "varIdentType";
    RuleName["setVarIdentType"] = "setVarIdentType";
    RuleName["identifierKeyword"] = "identifierKeyword";
    RuleName["identifierKeywordsAmbiguous1RolesAndLabels"] = "identifierKeywordsAmbiguous1RolesAndLabels";
    RuleName["identifierKeywordsAmbiguous2Labels"] = "identifierKeywordsAmbiguous2Labels";
    RuleName["labelKeyword"] = "labelKeyword";
    RuleName["identifierKeywordsAmbiguous3Roles"] = "identifierKeywordsAmbiguous3Roles";
    RuleName["identifierKeywordsUnambiguous"] = "identifierKeywordsUnambiguous";
    RuleName["roleKeyword"] = "roleKeyword";
    RuleName["lValueKeyword"] = "lValueKeyword";
    RuleName["identifierKeywordsAmbiguous4SystemVariables"] = "identifierKeywordsAmbiguous4SystemVariables";
    RuleName["roleOrIdentifierKeyword"] = "roleOrIdentifierKeyword";
    RuleName["roleOrLabelKeyword"] = "roleOrLabelKeyword";
})(RuleName = exports.RuleName || (exports.RuleName = {}));
